### ISSUE TITLE
`parser.hooks.new` not called for classes within the same file. 

### ISSUE THREAD

**Issue by Danielku15** [webpack/webpack#18113]
# Bug report

**What is the current behavior?**

The `parser.hooks.new("Item")` hooks are not called when the class `Item` is within the same file/module. It works if the class `Item` is imported from another module.

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Use this repository: https://github.com/Danielku15/webpack-parser-hook-call
2. Run `npm install && npm run build`
3. Notice in the console output that the hook is not called for `new NotWorkingClass`

In the `webpack.config.js` you will find a small plugin tapping into the `parser.hooks.new` for ESM javascript files. It then expects to be called for all the `new` expressions within input.mjs. But there is no call for `NotWorkingClass`

**What is the expected behavior?**

The parser hook should also be called for the `new NotWorkingClass` expression. From what I could see is following path prevents the hook to be called: 

* `JavaScriptParser.walkNewExpression` is called for all expressions which calls `JavaScriptParser.callHooksForExpression`
* For `WorkingClass` and `Uint8Array` the `JavaScriptParser.getMemberExpressionInfo` returns some data which allows the right hooks to be called. 
* For `NotWorkingClass` undefined is returned. This is because in `JavaScriptParser.getFreeInfoFromVariable` the `freeName` is set to `true` instead of a string bailing the whole code path:
https://github.com/webpack/webpack/blob/611bded369b5a9ea695f2669b25a6f88b67d7826/lib/javascript/JavascriptParser.js#L4535

I'm not sure what `getFreeInfoFromVariable` should return in this case from a low level behavior. Either the parser somehow needs to fill `freeName` to the className or the hook calling might need adoption to work with Identifiers.

**Other relevant information:**
webpack version: 5.90.3
Node.js version: v20.9.0
Operating System: Windows 11 Professional
Additional tools:

**Comment by Danielku15** [comment]
I found something suspicious which might allow an easy fix: 

1. Various top level symbols are marked by the InnerGraphPlugin: 
  * https://github.com/webpack/webpack/blob/611bded369b5a9ea695f2669b25a6f88b67d7826/lib/optimize/InnerGraphPlugin.js#L133
2. The `InnerGraph.tagTopLevelSymbol` first defines a variable in the parser with the name and then additionally tags the variable with the top level symbol. 
  * https://github.com/webpack/webpack/blob/611bded369b5a9ea695f2669b25a6f88b67d7826/lib/optimize/InnerGraph.js#L269-L279
3. The problem: 
  1. `parser.defineVariable` [adds a variable](https://github.com/webpack/webpack/blob/611bded369b5a9ea695f2669b25a6f88b67d7826/lib/javascript/JavascriptParser.js#L4385-L4391) with the current scope as value. 
  2. `parser.tagVariable` [also defines a variable](https://github.com/webpack/webpack/blob/611bded369b5a9ea695f2669b25a6f88b67d7826/lib/javascript/JavascriptParser.js#L4356-L4380) but with the correct top level symbol as value. But as there is already a VariableInfo registered from `defineVariable` [resulting in the name not to be set](https://github.com/webpack/webpack/blob/611bded369b5a9ea695f2669b25a6f88b67d7826/lib/javascript/JavascriptParser.js#L4373) 

I see two potential fixes in this code path: 

1. In `InnerGraph.tagTopLevelSymbol` we remove the call to `parser.defineVariable`. (or move it into the if-statement when a tag already exists). 
   * Feels like the safer option as the change is quite isolated and just ensures we do not double-define the variable with missing info. 
3. In `JavaScriptParser.tagVariable` we fill the name we know instead of `true` into the VariableInfo. 
   * It seems the old name was used before and things were refactored as part of https://github.com/webpack/webpack/commit/ec518945f18618d4c3b4992f129442e73dd331da I'm not so sure about the impact of this change.

**Comment by alexander-akait** [comment]
Feel free to send a PR, we have a lot of tests, so we will catch potential problems

**Comment by alexander-akait** [comment]
Just looked at the reproducible test repo, yeah, we need to fix it, because other plugins can be broken, it can work for development mode, but broken for the production mode, it is a critical issue... Will you look at this?

**Comment by Danielku15** [comment]
Yes, I will try to open a PR soon. Should be there the next few days latest. Locally solution 1 worked fine for me. But have to check all unit tests first.

**Comment by Danielku15** [comment]
Made a proposal for fixing the problem. I'm willing to contribute the change as an individual developer but I'm a bit reluctant to provide so many personal details (e.g. mailing address) in the EasyCLA to just contribute some code to this project.

**Comment by VivekShahare04** [comment]
I wanted to resolve this issue

**Comment by VivekShahare04** [comment]
JavaScriptParser.getFreeInfoFromVariable function sets freeName to true instead of a string for NotWorkingClass, causing the parser to bail out and skip the hook registration process.

**Comment by VivekShahare04** [comment]
see if we directly adjust the hook registration logic to work with identifiers ,
(node:8984) [DEP_WEBPACK_COMPILATION_NORMAL_MODULE_LOADER_HOOK] DeprecationWarning: Compilation.hooks.normalModuleLoader was moved to NormalModule.getCompilationHooks(compilation).loader

it will give you just deprecation warning ,so if incase we adjust it so that we're tapping into the expression hook for the "new" keyword. Then, we check if the argument to new is an identifier (assuming it's a simple class name without further qualification) and directly check if it matches "NotWorkingClass"
This approach circumvents the need for the parser to fill freeName correctly and directly handles the class name. Try this adjustment in your code to see if it resolves the issue with NotWorkingClass expressions not triggering the hook.

I am going to attach my piece of code......
![Screenshot (44)](https://github.com/webpack/webpack/assets/138312154/0c371c39-ccf0-4613-9a43-e24a39a9b35f)


tell me if it helps you if not then we will work on this issue more

**Comment by alexander-akait** [comment]
bump

**Comment by hai-x** [comment]
I think we should not rely on `innerGraph`, even if user set`innergraph` to false, `new` for `NotWorkingClass` also should be called.

**Comment by hai-x** [comment]
> @hai-x Want to try sending a PR with changes?

I'll look deeper and try to solve it.

**Comment by Sadaf-A** [comment]
@alexander-akait I would like solve this issue could you please assign?

**Comment by alexander-akait** [comment]
@Sadaf-A I don't assign, because it is an open source, just send a PR with a fix

**Comment by Ramjee194** [comment]
please assign it to me.I know the solution

![Image](https://github.com/user-attachments/assets/3a7abd8e-cc3f-47a0-9e27-00eb4629eeba)
![Image](https://github.com/user-attachments/assets/d8f44792-e774-4a46-bf16-a171721bdbfc)

**Comment by alexander-akait** [comment]
PR welcome

**Comment by anjith1** [comment]
Hi maintainers,

I'm K Anjith Goud, a 2nd-year Information Technology student from India. I'm exploring contributions for GSoC 2025 and I’m particularly interested in Webpack due to my strong background in JavaScript, Node.js, and web development.

While going through open issues, I found #18113 (https://github.com/webpack/webpack/issues/18113) very interesting. I've successfully reproduced the issue using the sample repo provided, and I'm diving into the internals of `JavascriptParser.js` to better understand the behavior.

From what I’ve observed, the root cause seems to be that `getFreeInfoFromVariable` returns `true` instead of the class name string when the class is defined in the same file. As a result, `getMemberExpressionInfo` fails to return the necessary data, which stops `callHooksForExpression` from executing the hook.

A possible fix could be updating the logic inside `getFreeInfoFromVariable` to identify local class declarations and return their name instead of `true`, or alternatively, modifying `callHooksForExpression` to handle `true` as a valid identifier in some special cases.

I’d love to hear your thoughts on whether this direction sounds correct or if you’d suggest a different approach. I’m excited to dive deeper and contribute meaningfully to the project.

Best regards,  
K ANJITH GOUD

**Comment by xiaoxiaojx** [comment]
This issue isn’t something that can be fixed by just changing a line or two, it’s more complex than it looks and involves a lot of logic. Some time will be needed to understand the original intent of the code.

**Comment by xiaoxiaojx** [comment]
Based on the [JavascriptParser.unittest.js#L115](https://github.com/webpack/webpack/blob/main/test/JavascriptParser.unittest.js#L115) here, this seems to be the expected behavior. As long as the variable is declared within the current scope (excluding undeclared globals or imported ones), its hooks.call won't be triggered, The behavior of hooks.new is consistent with this.

**Comment by ahabhgk** [comment]
Users have to tag the variable `tagVariable()` at preDeclarator hook (or any hook before innerGraph call `defineVariable()` in `tagTopLevelSymbol()`, yes the order matters here) if they want call hooks for the unfree/declared variable, an example is [WorkerPlugin](https://github.com/webpack/webpack/blob/feaa9339966416effed2ebb75a8a8ac206138018/lib/dependencies/WorkerPlugin.js#L492-L503)

**Comment by alexander-akait** [comment]
@ahabhgk Yeah, we use this logic before, so I think we can close this issue (because works as expected) and recommended to use this approach, what do you think?

**Comment by ahabhgk** [comment]
I agree

**Comment by alexander-akait** [comment]
Closed, works as expected, please use such solution - https://github.com/webpack/webpack/issues/18113#issuecomment-3210126748
